adapters/sspbc/sspbc.go

```diff
@@ -0,0 +1,407 @@
+	"github.com/mxmCherry/openrtb/v15/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const version = "5.6"
+
+// MC payload (for banner ads)
+type McAd struct {
+	Id      string             `json:"id"`
+	Seat    string             `json:"seat"`
+	SeatBid []openrtb2.SeatBid `json:"seatbid"`
+}
+
+// Adslot data (oneCode detection)
+type AdSlotData struct {
+	PbSlot string `json:"pbslot"`
+	PbSize string `json:"pbsize"`
+}
+
+// Banner Template payload
+type TemplatePayload struct {
+	SiteId  string `json:"siteid"`
+	SlotId  string `json:"slotid"`
+	AdLabel string `json:"adlabel"`
+	PubId   string `json:"pubid"`
+	Page    string `json:"page"`
+	Referer string `json:"referer"`
+	McAd    string `json:"mcad"`
+}
+
+// Ext data in request.imp
+type SsbcRequestImpExt struct {
+	Data AdSlotData `json:"data"`
+}
+
+// Ext data added by proxy
+type SsbcResponseExt struct {
+	AdLabel     string `json:"adlabel"`
+	PublisherId string `json:"pubid"`
+	SiteId      string `json:"siteid"`
+	SlotId      string `json:"slotid"`
+}
+
+type SspbcAdapter struct {
+	version  string
+	endpoint string
+	// adslots mapping
+	// map key is slot id (as sent and received from proxy)
+	adSlots        map[string]AdSlotData
+	adSizes        map[string]int
+	bannerTemplate *template.Template
+}
+
+// ---------------ADAPTER INTERFACE------------------
+// Builder builds a new instance of the sspBC adapter
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	// find path to template file
+	pathToTemplate := "./adapters/sspbc/bannerTemplate.html"
+	workingDir, err := os.Getwd()
+	if err == nil && strings.HasSuffix(workingDir, "sspbc") {
+		// this is a test running in adapter's directory
+		pathToTemplate = "bannerTemplate.html"
+	} else if err != nil {
+		// this error does not break createBannerAd flow
+		glog.Errorf("SSPBC: Cannot get working directory, assuming default path")
+	}
+
+	bannerTemplate, err := template.ParseFiles(pathToTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	bidder := &SspbcAdapter{
+		endpoint:       config.Endpoint,
+		version:        version,
+		bannerTemplate: bannerTemplate,
+	}
+
+	return bidder, nil
+}
+
+func (a *SspbcAdapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errors []error
+
+	formattedRequest, err := formatSsbcRequest(a, request)
+	if err != nil {
+		glog.Errorf("SSPBC: cannot prepare request")
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	requestJSON, err := json.Marshal(formattedRequest)
+	if err != nil {
+		glog.Errorf("SSPBC: cannot marshal request")
+		errors = append(errors, err)
+		return nil, errors
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    fmt.Sprintf("%s?bdver=%s&inver=0", a.endpoint, a.version),
```

Comment:
Consider using the `url` package to append query parameters instead of relying on string templating as that approach is more robust.

Reply:
Ok, changed tu url.